src/decode.rs

```diff
@@ -762,9 +762,13 @@ impl<C: CodeBuffer> Stateful for DecodeState<C> {
                 }
 
                 // Check that we don't overflow the code size with all codes we burst decode.
-                let potential_code = self.next_code + burst_size as u16;
-                burst_size += 1;
-                if potential_code == self.code_buffer.max_code() - Code::from(self.is_tiff) {
+                if let Some(potential_code) = self.next_code.checked_add(burst_size as u16) {
+                    burst_size += 1;
+                    if potential_code == self.code_buffer.max_code() - Code::from(self.is_tiff) {
+                        break;
+                    }
+                } else {
+                    // next_code overflowed
```

Comment:
Hm. That's an odd bug. This overflow would imply there's a code to be assigned larger than `max_code`. In particular we _will_ assign one later on. Since codes are at most [12bit](https://github.com/image-rs/lzw/blob/d55c610f1bff4ef0ed2f72af2c823b4a244d51a3/src/lib.rs#L64) and not 16, reaching the overflow should be impossible without incorrectly assumed codes before that point.

Reply:
The issue I was seeing was that this code was being called in a loop, and `next_code` just eventually stepped its way past 64K (it started around 64479 using the provided repro).

Reply:
```
[/.../weezl/src/decode.rs:765] &self.next_code = 18
[/.../weezl/src/decode.rs:765] &burst_size = 0
...
[/,,,/weezl/src/decode.rs:765] &self.next_code = 65535
[/.../weezl/src/decode.rs:765] &burst_size = 0
[/.../weezl/src/decode.rs:765] &self.next_code = 65535
[/.../src/decode.rs:765] &burst_size = 1
[/.../weezl/src/decode.rs:765] &self.next_code = 65535
[/.../Src/Rust/weezl/src/decode.rs:765] &burst_size = 0

Reply:
In the non-optimized code path the increment itself is gated behind `!self.table.is_full()` https://github.com/image-rs/lzw/blob/d55c610f1bff4ef0ed2f72af2c823b4a244d51a3/src/decode.rs#L893-L904 aka `self.inner.len() < MAX_ENTRIES` where `MAX_ENTRIES = 1 << 12`.

I'm decently sure that if we add a debug assertion to `Table::derive`, which asserts this invariant, then we happen across a much sooner violation.